### ISSUE TITLE
[🚀 Feature]: Need a way to enforce HTTP1.1 usage with the Java HTTP client

### ISSUE THREAD

**Issue by mykola-mokhnach** [SeleniumHQ/selenium#12918]
### Feature and motivation

Currently it is not possible to enforce HTTP1.1 usage in the client. This feature is needed because Appium server only supports HTTP1.1 so far (SPDY support has been added [recently](https://github.com/appium/appium/pull/19105), but it's optional), so we'd like to default to HTTP1.1 for good.

### Usage example

Not sure yet where exactly this needs to live.

**Comment by titusfortner** [comment]
I started working on this, looks straightforward, hope to have something we can put in nightly

**Comment by joerg1985** [comment]
I think this is nothing the client should do.

The inital request is done with http 1.1 in any case and will ony upgrade to http/2 when the server will respond with switch protocol.

The server should remove the http Upgrade header (not the one for WebSockets ;)) to avoid the http/2 upgrade.

**Comment by mykola-mokhnach** [comment]
> I think this is nothing the client should do.
> 
> The inital request is done with http 1.1 in any case and will ony upgrade to http/2 when the server will respond with switch protocol.
> 
> The server should remove the http Upgrade header (not the one for WebSockets ;)) to avoid the http/2 upgrade.

maybe. What I observe for now is that we did not see connection issues with previous client versions using netty @asolntsev got a similar results with their [tests](https://github.com/appium/java-client/pull/2036#issuecomment-1756859338). And enforcing HTTP version was the only workaround I was able to google. 
@joerg1985 I would be happy to accept any suggestions to avoid the aforementioned connection errors that don't require this particular change.

**Comment by joerg1985** [comment]
@mykola-mokhnach I just had a look at the initial request and it is like expected.
You can enable the logging by using `System.setProperty("jdk.httpclient.HttpClient.log", "headers,content");`

The only difference in the request are the 'Connection', 'HTTP2-Settings', 'Upgrade' headers.

HTTP2 enabled client:

```
POST /session HTTP/1.1
Connection: Upgrade, HTTP2-Settings
Content-Length: 215
Host: 127.0.0.1:5555
HTTP2-Settings: AAEAAEAAAAIAAAABAAMAAABkAAQBAAAAAAUAAEAA
Upgrade: h2c
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
User-Agent: selenium/4.14.0 (java windows)
```

HTTP2 disabled client:

```
POST /session HTTP/1.1
Content-Length: 215
Host: 127.0.0.1:5555
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
User-Agent: selenium/4.14.0 (java windows)
```

So i assume either the Appium http server or one of the drivers are not able to process this headers.
If the driver fails processing the headers the Appium http server could filter these headers.
The ReverseProxyHandler of the Selenium server does filter these and other headers.
Is this done by the Appium server too or are the headers just passed to the driver?

If you need to have a workaround for now, it might be possible to create a system property that selenium will read and force the http1.1 version. @diemol @titusfortner what do you think about this? this should be removed in selenium 5 and is only a workaround for this issue

**Comment by titusfortner** [comment]
As I see it:

1. Fix Appium http server
2. Have Java Appium require the usage of the JDK HTTP Client  specifically and force the version
3. System Property that toggles the version in the JDK HTTP Client that Appium can set as default
4. Add version to the Client Config and have the JDK client use it if found (#12919)

We should at least file an issue for the first one

I prefer 4 to 3 from the Selenium side.

**Comment by mykola-mokhnach** [comment]
> As I see it:
> 
> 1. Fix Appium http server
> 2. Have Java Appium require the usage of the JDK HTTP Client  specifically and force the version
> 3. System Property that toggles the version in the JDK HTTP Client that Appium can set as default
> 4. Add version to the Client Config and have the JDK client use it if found ([[java] allow setting version in the Http Client Config #12919](https://github.com/SeleniumHQ/selenium/pull/12919))
> 
> We should at least file an issue for the first one
> 
> I prefer 4 to 3 from the Selenium side.

`#1` can hardly be fixed now as ExpressJS that we use to serve HTTP does not have HTTP2 support and it is not very clear when it is going to have it.

**Comment by titusfortner** [comment]
Is there any consideration of switching off ExpressJS or using spdy package with it?

**Comment by mykola-mokhnach** [comment]
> switching off ExpressJS 

this would be a huge refactoring without any visible improvements and multiple breaking changes

> using spdy package with it

The last commit to spdy package happened 2 years ago. It has been incorporated recently, although is only used if server-side https is enabled. It's too much of risk to enable it by default for plain http. 

cc @jlipps

**Comment by joerg1985** [comment]
@mykola-mokhnach i do not think the appium server must support http/2 to fix this. The upgrade to http/2 is only done by the client when the server does answer with 
[switch protocol](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade) 
It is more about removing some headers from the request send to the Appium server before passing them to the drivers. The selenium server is doing this in the [ReverseProxyHandler](https://github.com/SeleniumHQ/selenium/blob/27c7fdbef8653ab1727548be76f5b8b53086462d/java/src/org/openqa/selenium/grid/web/ReverseProxyHandler.java).

So again the question, does the appium server allready remove these headers from the request before sending it to the driver?

**Comment by mykola-mokhnach** [comment]
If this is only about removing particular headers then I could try something like https://github.com/appium/java-client/pull/2036/commits/924838fdb53ee85ae39aad2660f086ef2b74ad99

**Comment by joerg1985** [comment]
These headers are added by the java http client automatically, there is no way to suppress this on the client side. These headers are part of the protocol and the http client does not allow to break the protocol.

The filter must be added to the server side, i think there is no other way to handle this.

**Comment by mykola-mokhnach** [comment]
Then I would rather use the solution provided by @titusfortner 
Let say I apply the changes to the server now:
- Existing versions of it still won't have them. I don't expect all users to upgrade their server after the client upgrade
- There is still no guarantee about the client behavior after this change is applied

**Comment by joerg1985** [comment]
I agree you need a workaround to this for some time in any case.

**Comment by titusfortner** [comment]
Is what was implemented in Selenium in 4.14.1 sufficient to close this?

**Comment by mykola-mokhnach** [comment]
yes, I would say it is ok now.
thanks for your efforts.
